### ISSUE TITLE
Panic when trying to sign

### ISSUE THREAD

**Issue by tonsV2** [lestrrat-go/jwx#389]
go version go1.16.3 linux/amd64

I have the following function
```bash
func generateAccessToken(user *model.User, key *rsa.PrivateKey, expiration int64) (string, error) {
	unixTime := time.Now().Unix()
	tokenExpiration := unixTime + expiration

	token := jwt.New()

	err := token.Set(jwt.IssuedAtKey, unixTime)
	if err != nil {
		return "", err
	}

	err = token.Set(jwt.ExpirationKey, tokenExpiration)
	if err != nil {
		return "", err
	}

	err = token.Set("user", user)
	if err != nil {
		return "", err
	}

	signed, err := jwt.Sign(token, jwa.RS256, key)
	if err != nil {
		log.Printf("Failed to sign token: %s", err)
		return "", err
	}

	return string(signed), nil
}
```

I generate my keys
```bash
openssl genpkey -algorithm RSA -out ./rsa_private.pem -pkeyopt rsa_keygen_bits:2048
openssl rsa -in ./rsa_private.pem -pubout -out ./rsa_public.pem
```

I test my function
```bash
func TestGenerateAccessToken(t *testing.T) {
	user := &model.User{
		Email:    "email",
		Password: "pass",
	}

	data, _ := ioutil.ReadFile("./../rsa_private.pem")

	decode, _ := pem.Decode(data)

	key, _ := x509.ParsePKCS1PrivateKey(decode.Bytes)

	token, _ := generateAccessToken(user, key, 12)

	println(token)
}
```

I get the following error
```bash
GOROOT=/snap/go/7360 #gosetup
GOPATH=/home/tons/go #gosetup
/snap/go/7360/bin/go test -c -o /tmp/___1TestGenerateAccessToken_in_github_com_dhis2_sre_instance_manager_service.test github.com/dhis2-sre/instance-manager/service #gosetup
/snap/go/7360/bin/go tool test2json -t /tmp/___1TestGenerateAccessToken_in_github_com_dhis2_sre_instance_manager_service.test -test.v -test.paniconexit0 -test.run ^\QTestGenerateAccessToken\E$
=== RUN   TestGenerateAccessToken
Ok!
--- FAIL: TestGenerateAccessToken (0.00s)
panic: reflect: call of reflect.Value.Type on zero Value [recovered]
	panic: reflect: call of reflect.Value.Type on zero Value

goroutine 6 [running]:
testing.tRunner.func1.2(0xa093a0, 0xc00000efa8)
	/snap/go/7360/src/testing/testing.go:1143 +0x332
testing.tRunner.func1(0xc0001aa600)
	/snap/go/7360/src/testing/testing.go:1146 +0x4b6
panic(0xa093a0, 0xc00000efa8)
	/snap/go/7360/src/runtime/panic.go:965 +0x1b9
reflect.Value.Type(0x0, 0x0, 0x0, 0x0, 0x0)
	/snap/go/7360/src/reflect/value.go:1908 +0x189
github.com/lestrrat-go/blackmagic.AssignIfCompatible(0xa4b280, 0xc0001d1800, 0xa4b280, 0x0, 0x0, 0x0)
	/home/tons/go/pkg/mod/github.com/lestrrat-go/blackmagic@v1.0.0/blackmagic.go:45 +0x20d
github.com/lestrrat-go/jwx/internal/keyconv.RSAPrivateKey(0xa4b280, 0xc0001d1800, 0xa4b280, 0x0, 0xc00009bac8, 0x509876)
	/home/tons/go/pkg/mod/github.com/lestrrat-go/jwx@v1.2.0/internal/keyconv/keyconv.go:36 +0xcd
github.com/lestrrat-go/jwx/jws.RSASigner.Sign(0xc0001fc510, 0xa8efda, 0x5, 0xc000236000, 0xf3, 0x2ae, 0xa4b280, 0x0, 0x509559, 0xc000236025, ...)
	/home/tons/go/pkg/mod/github.com/lestrrat-go/jwx@v1.2.0/jws/rsa.go:101 +0x79
github.com/lestrrat-go/jwx/jws.(*Signature).Sign(0xc00009bcf0, 0xc000232140, 0x9a, 0xa0, 0xb2b950, 0xc00000ef48, 0xa4b280, 0x0, 0x0, 0x0, ...)
	/home/tons/go/pkg/mod/github.com/lestrrat-go/jwx@v1.2.0/jws/message.go:84 +0x3f0
github.com/lestrrat-go/jwx/jws.Sign(0xc000232140, 0x9a, 0xa0, 0xa8efda, 0x5, 0xa4b280, 0x0, 0xc00009be18, 0x1, 0x1, ...)
	/home/tons/go/pkg/mod/github.com/lestrrat-go/jwx@v1.2.0/jws/jws.go:98 +0x1e9
github.com/lestrrat-go/jwx/jwt.Sign(0xb35348, 0xc0001d15c0, 0xa8efda, 0x5, 0xa4b280, 0x0, 0x0, 0x0, 0x0, 0xc0000ba900, ...)
	/home/tons/go/pkg/mod/github.com/lestrrat-go/jwx@v1.2.0/jwt/jwt.go:267 +0x2c6
github.com/dhis2-sre/instance-manager/service.generateAccessToken(0xc0000ba880, 0x0, 0xc, 0x0, 0xb27380, 0xc0001c78e0, 0x1)
	/home/tons/IdeaProjects/dhis2/hosting-platform/service/tokenHelper.go:34 +0x26c
github.com/dhis2-sre/instance-manager/service.TestGenerateAccessToken(0xc0001aa600)
	/home/tons/IdeaProjects/dhis2/hosting-platform/service/tokenHelper_test.go:29 +0x171
testing.tRunner(0xc0001aa600, 0xab4b38)
	/snap/go/7360/src/testing/testing.go:1193 +0xef
created by testing.(*T).Run
	/snap/go/7360/src/testing/testing.go:1238 +0x2b3


Process finished with the exit code 1
```

Any clue what I'm doing wrong?

**Comment by floodkoff** [comment]
```
	var privkey rsa.PrivateKey
	if err := keyconv.RSAPrivateKey(&privkey, key); err != nil {
		return nil, errors.Wrapf(err, `failed to retrieve rsa.PrivateKey out of %T`, key)
	}
```
Looking at the code that paniced. I'm curious what is the difference from:
```
    privKey, ok := key.(*rsa.PrivateKey)
    if !ok {
        return nil, errors.Wrapf(err, `invalid key type`, key)
    }
```

**Comment by lestrrat** [comment]
Still investigating, but meanwhile, you should be able to read PEM encoded keys to a JWK, and pass that to any of the methods (untested):

```go
key, err := jwk.ReadFile(`../rsa_private.pem`, jwk.WithPEM(true))
... snip ...
signed, err := jws.Sign(token, jwa.RS256, key)
```

**Comment by lestrrat** [comment]
Okay, I got it. You wrote out your key in  PKCS8 format. But you read it as PKCS1.
And you *should* have caught the error at

```go
	key, _ := x509.ParsePKCS1PrivateKey(decode.Bytes)
```

which, at least in your snippet, you did not check the error for. If you actually checked for errors, you shouldn't have encountered this error.

====

Now, why is it panic'ing? Am I not checking for `key == nil`? You bet I am.

The tricky bit is from here. The invocation for `generateAccessToken()` is receiving `key = nil`, but the `key` parameter is typed as `*rsa.PrivateKey`. However, `jwt.Sign()` takes the key as `interface{}`, so all functions `jwt.Sign()` onwards are receiving something that looks like this (not exact, but just trying to show the general idea): `interface{}(*rsa.PrivateKey(nil))`

This is what's known as a "typed nil" in Go.

For example,

```go
type CustomError struct {}
// assume we have all the necessary bits for CustomError to fulfill the `error` interface

func Foo() error {
   var err *CustomError
   // err is "typed" as *CustomError, but it's nil
   return err // ... and we just return it
}

func main() {
   err := Foo()
   fmt.Println(err == nil) // <--- HERE
}
```

I haven't actually run the code, but trust me on this: the line "HERE" prints `false`.


So I went out of my way to explain the above because this is what I wanted to say:
I do check for `key == nil`, but the above slips through the cracks.
I do believe there are ways to detect this, but I find it to be not worth the trouble, because in most cases doing the following almost always avoids this:

1. *Always* check for errors (this is where you messed up)
2. *Always* return an explicit nil when errors are encountered (i.e. `return nil`)

HTH
